File: modules/data/tgate.lua
sptbl["tgate"] = {

    files = {
        module = "tgate.c",
        header = "tgate.h",
        example = "ex_tgate.c",
    },

    func = {
        create = "ut_tgate_create",
        destroy = "ut_tgate_destroy",
        init = "ut_tgate_init",
        compute = "ut_tgate_compute",
    },

    params = {
        optional = {
            {
                name = "time",
                type = "UTFLOAT",
                description = "Duration of the gate (in seconds)",
                default = 0.5
            },
        }
    },

    modtype = "module",

    description = [[A triggerable gate.
]],

    ninputs = 1,
    noutputs = 1,

    inputs = {
        {
            name = "trigger",
            description = "When non-zero, will (re)trigger gate."
        },
    },

    outputs = {
        {
            name = "gate",
            description = "Gate output."
        },
    }

}

File: modules/tgate.c
#include <stdlib.h>
#include "utone.h"

int ut_tgate_create(ut_tgate **p)
{
    *p = malloc(sizeof(ut_tgate));
    return UT_OK;
}

int ut_tgate_destroy(ut_tgate **p)
{
    free(*p);
    return UT_OK;
}

int ut_tgate_init(ut_data *ut, ut_tgate *p)
{
    p->time = 0;
    p->timer = 0;
    return UT_OK;
}

int ut_tgate_compute(ut_data *ut, ut_tgate *p, UTFLOAT *in, UTFLOAT *out)
{
    *out = 0;
    if(*in != 0) {
        p->timer = p->time * ut->sr;
    }
    if(p->timer != 0) {
        *out = 1;
        p->timer--;
    }
    return UT_OK;
}
